server/src/server.ts

```diff
@@ -918,71 +873,237 @@ function getQuickFix(diagnostic: Diagnostic, title: string, range: Range, replac
+		const isStruct = line.startsWith(STRUCT) && line.length > STRUCT.length+1 && line.charAt(STRUCT.length).match(/\s/);
+		const isNamespace = line.startsWith(NAMESPACE) && line.length > NAMESPACE.length+1 && line.charAt(NAMESPACE.length).match(/\s/);
+		if (isFunction || isStruct || isNamespace) {
+			let importItem = line.split(/[\s{(:]+/)[1];
+			importItems.add(importItem);
+		}				
+	}
+
+	return importItems;
+}
+
+/**
+ * @summary Gets text before and after the cursor ( on the same line )
+ * @param position 
+ * @param textDocumentFromURI 
+ * @returns Text before and after the cursor position | undefined
+ */
+function getTextAroundCursor(position: Position, textDocumentFromURI: TextDocument) {
+	const startOfLine = { 
+		line: position.line, 
+		character: 0 
+	};
+	const cursorPosition = { 
+		line: position.line, 
+		character: position.character
+	};
+	const nextLine = {
+		line: position.line + 1,
+		character: 0,
+	};
+
+	const lineUpToCursor = textDocumentFromURI.getText({ start: startOfLine, end: cursorPosition });
+	const lineUpToCursorSplit = lineUpToCursor.split(/\s+/);
+	if(lineUpToCursorSplit?.[0] === undefined)
+		return undefined
+
+	const textBeforeCursor = lineUpToCursorSplit[lineUpToCursorSplit.length - 1];
+		
+	const textAfterCursor = textDocumentFromURI
+		.getText({ start: cursorPosition, end: nextLine })?.split(/\s+/)[0];
+	
+	return { textBeforeCursor, textAfterCursor };
+}
+
+/**
+ * @summary Infers the type of the syntax the cursor is pointing at
+ * @param position 
+ * @param textDocumentFromURI 
+ * @returns SyntaxType of the current location
+ */
+function getSyntaxType(position: Position, textDocumentFromURI: TextDocument): SyntaxType {
+	const fileStart = { line: 0, character: 0 };
+	const cursorPosition = { line: position.line, character: position.character };
+
+	// Get text from the start of the document to position of the cursor
+	// This might be too exhaustive on a large file, should change if there's a better way to 
+	// determine if we're inside a function/from statement etc.
+	const textUpToCursor = textDocumentFromURI.getText({start: fileStart, end: cursorPosition});
+	
+	// The order here might matter, as they are evaluated top-to-bottom
+	const regexes = [
+		// func ...{}()
+		// Match func only if there's no end afterwards
+		{ exp: /^func(?!(.|\s)*^end)/gm, syn: SyntaxType.Function },
+		// from module import (
+		// Match only if there's no matching ")" meaning we're inside the parantheses
+		// After the parantheses it will also match any number of function names
+		{ exp: /^from[ \t]+[a-zA-Z0-9_.]*[ \t]+import[ \t]*\((?:[ \n\t]*[a-zA-Z0-9_.]*[ \t]*,*)*(?![\s\S]*\))/gm, syn: SyntaxType.ImportFunctionP },
+		// from module import
+		// Any number for function names seperated by commas are also matched
+		{ exp: /^from[ \t]+[a-zA-Z0-9_.]*[ \t]+import[ \t]*(?:[a-zA-Z0-9_.]*[ \t]*,*[ \t]*)*$/gm, syn: SyntaxType.ImportFunction },
```

Comment:
If the cursor is on an empty line (for example [this line](https://github.com/OpenZeppelin/cairo-contracts/blob/dfafdd2ebd778718cf440555e975f6eace278695/src/openzeppelin/token/erc721/ERC721_Mintable_Burnable.cairo#L8)), but the previous line was a complete `from` statement, then the `ImportFunction` regex would match, even though this scenario should really be the base case.  

Is the multiline mode really needed?  Perhaps avoiding multiline mode and just matching contents up to the end of the string is enough, since the string ends where the cursor is.

Reply:
I'll switch to iterative functions instead of regexes, that way I can be more specific about where I assume we are in which statement